velox/exec/Window.h

```diff
@@ -0,0 +1,228 @@
+  // used for sorting by those key combinations during the processing.
+  // partitionKeyInfo_ is used to separate partitions in the rows.
+  // sortKeyInfo_ is used to identify peer rows in a partition.
+  // allKeyInfo_ is a combination of (partitionKeyInfo_ and sortKeyInfo_).
+  // It is used to perform a full sorting of the input rows to be able to
+  // separate partitions and sort the rows in it. The rows are output in
+  // this order by the operator.
+  std::vector<std::pair<column_index_t, core::SortOrder>> partitionKeyInfo_;
+  std::vector<std::pair<column_index_t, core::SortOrder>> sortKeyInfo_;
+  std::vector<std::pair<column_index_t, core::SortOrder>> allKeyInfo_;
+
+  // Vector of WindowFunction objects required by this operator.
+  // WindowFunction is the base API implemented by all the window functions.
+  // The functions are ordered by their positions in the output columns.
+  std::vector<std::unique_ptr<exec::WindowFunction>> windowFunctions_;
+  // Vector of WindowFrames corresponding to each windowFunction above.
+  // It represents the frame spec for the function computation.
+  std::vector<WindowFrame> windowFrames_;
+
+  // This SelectivityVector is used across addInput calls for decoding.
+  SelectivityVector inputRows_;
+  // Number of input rows.
+  vector_size_t numRows_ = 0;
+
+  // Vector of pointers to each input row in the data_ RowContainer.
+  // The rows are sorted by partitionKeys + sortKeys. This total
+  // ordering can be used to split partitions (with the correct
+  // order by) for the processing.
+  std::vector<char*> sortedRows_;
+
+  // Window partition object used to provide per-partition
+  // data to the window function.
+  std::unique_ptr<WindowPartition> windowPartition_;
+
+  // Number of rows that be fit into an output block.
+  vector_size_t numRowsPerOutput_;
+
+  // This is a vector that gives the index of the start row of
+  // each partition in the RowContainer data_. The sorting follows
+  // the order in sortedRows_. This auxiliary structure helps
+  // demarcate partitions in getOutput calls.
+  std::vector<vector_size_t> partitionStartRows_;
+
+  // The following 4 Buffers are used to pass peer and frame start and
+  // end values to the WindowFunction::apply method. These
+  // buffers can be allocated once and reused across all the getOutput
+  // calls.
+  // Only a single peer start and peer end buffer is needed across all
+  // functions (as the peer values are based on the ORDER BY clause).
+  BufferPtr peerStartBuffer_;
+  BufferPtr peerEndBuffer_;
+  // A separate BufferPtr is required for the frame indexes of each
+  // function. Each function has its own frame clause and style. So we
+  // have as many buffers as the number of functions.
+  std::vector<BufferPtr> allFuncsFrameStartBuffer_;
+  std::vector<BufferPtr> allFuncsFrameEndBuffer_;
+
+  // The 4 below are for the raw pointers to the above 4 BufferPtrs.
```

Comment:
Why do we need these as member variables?

Reply:
I was trying to avoid writing code like
auto rawFrameStartBuffer = frameStartBuffer->asMutable<vector_size_t>();
auto rawFrameEndBuffer = frameEndBuffer->asMutable<vector_size_t>();

for each output buffer. wdyt ?

Reply:
These seem to be used only in two methods. So should be okay to write this code?

Reply:
I agree. It should be fine to remove these variables and use asMutable() when needed.